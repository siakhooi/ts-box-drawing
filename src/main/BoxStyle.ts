export class BoxStyle {
  HORIZONTAL: string;
  VERTICAL: string;
  TOP_LEFT: string;
  TOP_RIGHT: string;
  BOTTOM_LEFT: string;
  BOTTOM_RIGHT: string;
  CROSS: string;
  TOP_MIDDLE: string;
  BOTTOM_MIDDLE: string;
  MIDDLE_LEFT: string;
  MIDDLE_RIGHT: string;

  constructor(
    horizontal: string,
    vertical: string,
    top_left: string,
    top_right: string,
    bottom_left: string,
    bottom_right: string,
    cross: string,
    top_middle: string,
    bottom_middle: string,
    middle_left: string,
    middle_right: string
  ) {
    this.HORIZONTAL = horizontal;
    this.VERTICAL = vertical;
    this.TOP_LEFT = top_left;
    this.TOP_RIGHT = top_right;
    this.BOTTOM_LEFT = bottom_left;
    this.BOTTOM_RIGHT = bottom_right;
    this.CROSS = cross;
    this.TOP_MIDDLE = top_middle;
    this.BOTTOM_MIDDLE = bottom_middle;
    this.MIDDLE_LEFT = middle_left;
    this.MIDDLE_RIGHT = middle_right;
  }
}
