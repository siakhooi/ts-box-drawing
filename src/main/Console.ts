import util from 'util';

class ConsoleClass {
  println(format: string, ...texts: string[]): void {
    console.log(util.format(format, ...texts));
  }
}

export const Console = new ConsoleClass();
