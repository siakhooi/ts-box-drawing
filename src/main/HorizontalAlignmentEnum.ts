export enum HorizontalAlignmentEnum {
  LEFT,
  CENTER,
  RIGHT,
}
