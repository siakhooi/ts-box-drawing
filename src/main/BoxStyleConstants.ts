import {BoxStyle} from './BoxStyle';

const THIN = new BoxStyle(
  '\u2500',
  '\u2502',
  '\u250c',
  '\u2510',
  '\u2514',
  '\u2518'
);
const THICK = new BoxStyle(
  '\u2501',
  '\u2503',
  '\u250F',
  '\u2513',
  '\u2517',
  '\u251B'
);
const DOUBLE = new BoxStyle(
  '\u2550',
  '\u2551',
  '\u2554',
  '\u2557',
  '\u255A',
  '\u255D'
);
const THINCURVE = new BoxStyle(
  '\u2500',
  '\u2502',
  '\u256D',
  '\u256E',
  '\u2570',
  '\u256F'
);

export {THIN, THICK, DOUBLE, THINCURVE};
