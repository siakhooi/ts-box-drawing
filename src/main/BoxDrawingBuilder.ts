import {BoxStyle} from './BoxStyle';
import {BoxStyles} from './BoxStyles';
import {Console} from './Console';
import {Padding} from './Padding';

type BoxDataType = null | string | string[][];

function getColumnWidths(
  rowCount: number,
  columnCount: number,
  data: string[][]
): number[] {
  const columnWidths: number[] = [];
  for (let i = 0; i < columnCount; i++) {
    columnWidths[i] = 0;
    for (let j = 0; j < rowCount; j++) {
      if (data[j][i] !== undefined) {
        columnWidths[i] = Math.max(columnWidths[i], data[j][i].length);
      }
    }
  }
  return columnWidths;
}

function convertDataToArray(data: BoxDataType): string[][] {
  if (data === '' || data === null) return [];
  return typeof data === 'string' ? [[data]] : data;
}

export class BoxDrawingBuilder {
  private data: BoxDataType = [];
  private style = BoxStyles.THIN;

  private paddingLeft = new Padding();
  private paddingRight = new Padding();

  setData(data: BoxDataType): BoxDrawingBuilder {
    this.data = data;
    return this;
  }
  setStyle(style: BoxStyle): BoxDrawingBuilder {
    this.style = style;
    return this;
  }
  setDefaultPadLeft(number_of_spaces: number): BoxDrawingBuilder {
    this.paddingLeft.setDefault(number_of_spaces);
    return this;
  }
  setDefaultPadRight(number_of_spaces: number): BoxDrawingBuilder {
    this.paddingRight.setDefault(number_of_spaces);
    return this;
  }
  padLeftAll(numbers_of_spaces: number[]): BoxDrawingBuilder {
    this.paddingLeft.padAll(numbers_of_spaces);
    return this;
  }
  padLeft(column: number, number_of_spaces: number): BoxDrawingBuilder {
    this.paddingLeft.setPadding(column, number_of_spaces);
    return this;
  }
  padRight(column: number, number_of_spaces: number): BoxDrawingBuilder {
    this.paddingRight.setPadding(column, number_of_spaces);
    return this;
  }
  pad(
    column: number,
    number_of_spaces_on_left: number,
    number_of_spaces_on_right: number
  ): BoxDrawingBuilder {
    this.padLeft(column, number_of_spaces_on_left);
    this.padRight(column, number_of_spaces_on_right);
    return this;
  }
  private getHorizontalLines(
    columnCount: number,
    columnWidths: number[]
  ): string[] {
    let lineTop = '',
      lineMiddle = '',
      lineBottom = '';
    const boxStyle = this.style;

    lineTop = boxStyle.TOP_LEFT;
    lineMiddle = boxStyle.MIDDLE_LEFT;
    lineBottom = boxStyle.BOTTOM_LEFT;

    for (let i = 0; i < columnCount; i++) {
      const padLeft = this.paddingLeft.getPad(i);
      const padRight = this.paddingRight.getPad(i);
      const columnWidth = columnWidths[i] + padLeft + padRight;
      lineTop += boxStyle.HORIZONTAL_OUTER.repeat(columnWidth);
      lineMiddle += boxStyle.HORIZONTAL_INNER.repeat(columnWidth);
      lineBottom += boxStyle.HORIZONTAL_OUTER.repeat(columnWidth);

      if (i < columnCount - 1) {
        lineTop += boxStyle.TOP_MIDDLE;
        lineMiddle += boxStyle.CROSS;
        lineBottom += boxStyle.BOTTOM_MIDDLE;
      }
    }
    lineTop += boxStyle.TOP_RIGHT;
    lineMiddle += boxStyle.MIDDLE_RIGHT;
    lineBottom += boxStyle.BOTTOM_RIGHT;

    return [lineTop, lineMiddle, lineBottom];
  }
  private getDataLine(
    columnCount: number,
    columnWidths: number[],
    rowData: string[]
  ): string {
    const boxStyle = this.style;
    let s = boxStyle.VERTICAL_OUTER;
    for (let i = 0; i < columnCount; i++) {
      const padLeft = this.paddingLeft.getPad(i);
      const padRight = this.paddingRight.getPad(i);
      const cellData = rowData[i] === undefined ? '' : rowData[i];
      s += ' '.repeat(padLeft);
      s += cellData.padEnd(columnWidths[i], ' ');
      s += ' '.repeat(padRight);
      if (i < columnCount - 1) s += boxStyle.VERTICAL_INNER;
    }
    s += boxStyle.VERTICAL_OUTER;

    return s;
  }
  drawBox() {
    const data: string[][] = convertDataToArray(this.data);
    const rowCount = data.length;
    const columnCount = data.reduce((r, n) => Math.max(r, n.length), 0);

    const columnWidths: number[] = getColumnWidths(rowCount, columnCount, data);

    let lineTop = '',
      lineMiddle = '',
      lineBottom = '';
    [lineTop, lineMiddle, lineBottom] = this.getHorizontalLines(
      columnCount,
      columnWidths
    );

    Console.println(lineTop);

    for (let j = 0; j < rowCount; j++) {
      Console.println(this.getDataLine(columnCount, columnWidths, data[j]));
      if (j < rowCount - 1) Console.println(lineMiddle);
    }

    Console.println(lineBottom);
  }
}
