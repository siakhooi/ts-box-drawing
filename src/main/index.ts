import {BoxStyle} from './BoxStyle';
import {Console} from './Console';
import {THIN, THICK, DOUBLE, THINCURVE} from './BoxStyleConstants';

function drawBox(boxStyle: BoxStyle, text: string) {
  const w = text.length;
  Console.printf(
    '%s%s%s',
    boxStyle.TOP_LEFT,
    boxStyle.HORIZONTAL.repeat(w),
    boxStyle.TOP_RIGHT
  );
  Console.printf('%s%s%s', boxStyle.VERTICAL, text, boxStyle.VERTICAL);
  Console.printf(
    '%s%s%s',
    boxStyle.BOTTOM_LEFT,
    boxStyle.HORIZONTAL.repeat(w),
    boxStyle.BOTTOM_RIGHT
  );
}
export function drawBoxThin(text: string) {
  drawBox(THIN, text);
}
export function drawBoxThick(text: string) {
  drawBox(THICK, text);
}
export function drawBoxDouble(text: string) {
  drawBox(DOUBLE, text);
}
export function drawBoxThinCurve(text: string) {
  drawBox(THINCURVE, text);
}
