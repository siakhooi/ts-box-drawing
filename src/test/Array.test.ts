import {Console} from '../main/Console';
import {ALL_FUNCTIONS, ARRAY_TEST_DATA, TestDataUnit} from './TestData';

describe.each(ALL_FUNCTIONS)('Array', (drawBoxFunction: Function) => {
  test.each(ARRAY_TEST_DATA)(
    `${drawBoxFunction.name} [$title]`,
    (testData: TestDataUnit) => {
      const allData: string[] = [];
      Console.println = jest
        .fn()
        .mockImplementation(text => allData.push(text));

      drawBoxFunction(testData.data);
      expect(allData).toMatchSnapshot();
    }
  );
});
