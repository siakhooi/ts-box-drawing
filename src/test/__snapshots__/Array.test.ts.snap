// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array drawBoxDouble [Simple 1X1] 1`] = `
[
  "╔════════════╗",
  "║In 1x1 array║",
  "╚════════════╝",
]
`;

exports[`Array drawBoxDouble [Simple 2x2] 1`] = `
[
  "╔════════╦════════╗",
  "║Cell 1-1║Cell 2-1║",
  "╠════════╬════════╣",
  "║Cell 1-2║Cell 2-2║",
  "╚════════╩════════╝",
]
`;

exports[`Array drawBoxDouble [Variable Row Lengths 1+1+3] 1`] = `
[
  "╔═════════════════╦════════════╦═════════════╗",
  "║                 ║Cell 2-1 ABC║             ║",
  "╠═════════════════╬════════════╬═════════════╣",
  "║Cell 1-2 Avatar  ║            ║             ║",
  "╠═════════════════╬════════════╬═════════════╣",
  "║Cell 1-3 Lemmings║Cell 2-3 A  ║Cell 3-3 Bata║",
  "╚═════════════════╩════════════╩═════════════╝",
]
`;

exports[`Array drawBoxDouble [Variable Row Lengths 1+2] 1`] = `
[
  "╔════════╦════════╗",
  "║Cell 1-1║        ║",
  "╠════════╬════════╣",
  "║Cell 1-2║Cell 2-2║",
  "╚════════╩════════╝",
]
`;

exports[`Array drawBoxHorizontalDoubleVerticalThin [Simple 1X1] 1`] = `
[
  "╒════════════╕",
  "│In 1x1 array│",
  "╘════════════╛",
]
`;

exports[`Array drawBoxHorizontalDoubleVerticalThin [Simple 2x2] 1`] = `
[
  "╒════════╤════════╕",
  "│Cell 1-1│Cell 2-1│",
  "╞════════╪════════╡",
  "│Cell 1-2│Cell 2-2│",
  "╘════════╧════════╛",
]
`;

exports[`Array drawBoxHorizontalDoubleVerticalThin [Variable Row Lengths 1+1+3] 1`] = `
[
  "╒═════════════════╤════════════╤═════════════╕",
  "│                 │Cell 2-1 ABC│             │",
  "╞═════════════════╪════════════╪═════════════╡",
  "│Cell 1-2 Avatar  │            │             │",
  "╞═════════════════╪════════════╪═════════════╡",
  "│Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata│",
  "╘═════════════════╧════════════╧═════════════╛",
]
`;

exports[`Array drawBoxHorizontalDoubleVerticalThin [Variable Row Lengths 1+2] 1`] = `
[
  "╒════════╤════════╕",
  "│Cell 1-1│        │",
  "╞════════╪════════╡",
  "│Cell 1-2│Cell 2-2│",
  "╘════════╧════════╛",
]
`;

exports[`Array drawBoxHorizontalThickVerticalThin [Simple 1X1] 1`] = `
[
  "┍━━━━━━━━━━━━┑",
  "│In 1x1 array│",
  "┕━━━━━━━━━━━━┙",
]
`;

exports[`Array drawBoxHorizontalThickVerticalThin [Simple 2x2] 1`] = `
[
  "┍━━━━━━━━┯━━━━━━━━┑",
  "│Cell 1-1│Cell 2-1│",
  "┝━━━━━━━━┿━━━━━━━━┥",
  "│Cell 1-2│Cell 2-2│",
  "┕━━━━━━━━┷━━━━━━━━┙",
]
`;

exports[`Array drawBoxHorizontalThickVerticalThin [Variable Row Lengths 1+1+3] 1`] = `
[
  "┍━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━┑",
  "│                 │Cell 2-1 ABC│             │",
  "┝━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━┿━━━━━━━━━━━━━┥",
  "│Cell 1-2 Avatar  │            │             │",
  "┝━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━┿━━━━━━━━━━━━━┥",
  "│Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata│",
  "┕━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━┙",
]
`;

exports[`Array drawBoxHorizontalThickVerticalThin [Variable Row Lengths 1+2] 1`] = `
[
  "┍━━━━━━━━┯━━━━━━━━┑",
  "│Cell 1-1│        │",
  "┝━━━━━━━━┿━━━━━━━━┥",
  "│Cell 1-2│Cell 2-2│",
  "┕━━━━━━━━┷━━━━━━━━┙",
]
`;

exports[`Array drawBoxHorizontalThinVerticalDouble [Simple 1X1] 1`] = `
[
  "╓────────────╖",
  "║In 1x1 array║",
  "╙────────────╜",
]
`;

exports[`Array drawBoxHorizontalThinVerticalDouble [Simple 2x2] 1`] = `
[
  "╓────────╥────────╖",
  "║Cell 1-1║Cell 2-1║",
  "╟────────╫────────╢",
  "║Cell 1-2║Cell 2-2║",
  "╙────────╨────────╜",
]
`;

exports[`Array drawBoxHorizontalThinVerticalDouble [Variable Row Lengths 1+1+3] 1`] = `
[
  "╓─────────────────╥────────────╥─────────────╖",
  "║                 ║Cell 2-1 ABC║             ║",
  "╟─────────────────╫────────────╫─────────────╢",
  "║Cell 1-2 Avatar  ║            ║             ║",
  "╟─────────────────╫────────────╫─────────────╢",
  "║Cell 1-3 Lemmings║Cell 2-3 A  ║Cell 3-3 Bata║",
  "╙─────────────────╨────────────╨─────────────╜",
]
`;

exports[`Array drawBoxHorizontalThinVerticalDouble [Variable Row Lengths 1+2] 1`] = `
[
  "╓────────╥────────╖",
  "║Cell 1-1║        ║",
  "╟────────╫────────╢",
  "║Cell 1-2║Cell 2-2║",
  "╙────────╨────────╜",
]
`;

exports[`Array drawBoxHorizontalThinVerticalThick [Simple 1X1] 1`] = `
[
  "┎────────────┒",
  "┃In 1x1 array┃",
  "┖────────────┚",
]
`;

exports[`Array drawBoxHorizontalThinVerticalThick [Simple 2x2] 1`] = `
[
  "┎────────┰────────┒",
  "┃Cell 1-1┃Cell 2-1┃",
  "┠────────╂────────┨",
  "┃Cell 1-2┃Cell 2-2┃",
  "┖────────┸────────┚",
]
`;

exports[`Array drawBoxHorizontalThinVerticalThick [Variable Row Lengths 1+1+3] 1`] = `
[
  "┎─────────────────┰────────────┰─────────────┒",
  "┃                 ┃Cell 2-1 ABC┃             ┃",
  "┠─────────────────╂────────────╂─────────────┨",
  "┃Cell 1-2 Avatar  ┃            ┃             ┃",
  "┠─────────────────╂────────────╂─────────────┨",
  "┃Cell 1-3 Lemmings┃Cell 2-3 A  ┃Cell 3-3 Bata┃",
  "┖─────────────────┸────────────┸─────────────┚",
]
`;

exports[`Array drawBoxHorizontalThinVerticalThick [Variable Row Lengths 1+2] 1`] = `
[
  "┎────────┰────────┒",
  "┃Cell 1-1┃        ┃",
  "┠────────╂────────┨",
  "┃Cell 1-2┃Cell 2-2┃",
  "┖────────┸────────┚",
]
`;

exports[`Array drawBoxInnerDoubleOuterThin [Simple 1X1] 1`] = `
[
  "┌────────────┐",
  "│In 1x1 array│",
  "└────────────┘",
]
`;

exports[`Array drawBoxInnerDoubleOuterThin [Simple 2x2] 1`] = `
[
  "┌────────╥────────┐",
  "│Cell 1-1║Cell 2-1│",
  "╞════════╬════════╡",
  "│Cell 1-2║Cell 2-2│",
  "└────────╨────────┘",
]
`;

exports[`Array drawBoxInnerDoubleOuterThin [Variable Row Lengths 1+1+3] 1`] = `
[
  "┌─────────────────╥────────────╥─────────────┐",
  "│                 ║Cell 2-1 ABC║             │",
  "╞═════════════════╬════════════╬═════════════╡",
  "│Cell 1-2 Avatar  ║            ║             │",
  "╞═════════════════╬════════════╬═════════════╡",
  "│Cell 1-3 Lemmings║Cell 2-3 A  ║Cell 3-3 Bata│",
  "└─────────────────╨────────────╨─────────────┘",
]
`;

exports[`Array drawBoxInnerDoubleOuterThin [Variable Row Lengths 1+2] 1`] = `
[
  "┌────────╥────────┐",
  "│Cell 1-1║        │",
  "╞════════╬════════╡",
  "│Cell 1-2║Cell 2-2│",
  "└────────╨────────┘",
]
`;

exports[`Array drawBoxInnerDoubleOuterThinCurve [Simple 1X1] 1`] = `
[
  "╭────────────╮",
  "│In 1x1 array│",
  "╰────────────╯",
]
`;

exports[`Array drawBoxInnerDoubleOuterThinCurve [Simple 2x2] 1`] = `
[
  "╭────────╥────────╮",
  "│Cell 1-1║Cell 2-1│",
  "╞════════╬════════╡",
  "│Cell 1-2║Cell 2-2│",
  "╰────────╨────────╯",
]
`;

exports[`Array drawBoxInnerDoubleOuterThinCurve [Variable Row Lengths 1+1+3] 1`] = `
[
  "╭─────────────────╥────────────╥─────────────╮",
  "│                 ║Cell 2-1 ABC║             │",
  "╞═════════════════╬════════════╬═════════════╡",
  "│Cell 1-2 Avatar  ║            ║             │",
  "╞═════════════════╬════════════╬═════════════╡",
  "│Cell 1-3 Lemmings║Cell 2-3 A  ║Cell 3-3 Bata│",
  "╰─────────────────╨────────────╨─────────────╯",
]
`;

exports[`Array drawBoxInnerDoubleOuterThinCurve [Variable Row Lengths 1+2] 1`] = `
[
  "╭────────╥────────╮",
  "│Cell 1-1║        │",
  "╞════════╬════════╡",
  "│Cell 1-2║Cell 2-2│",
  "╰────────╨────────╯",
]
`;

exports[`Array drawBoxInnerThickOuterThin [Simple 1X1] 1`] = `
[
  "┌────────────┐",
  "│In 1x1 array│",
  "└────────────┘",
]
`;

exports[`Array drawBoxInnerThickOuterThin [Simple 2x2] 1`] = `
[
  "┌────────┰────────┐",
  "│Cell 1-1┃Cell 2-1│",
  "┝━━━━━━━━╋━━━━━━━━┥",
  "│Cell 1-2┃Cell 2-2│",
  "└────────┸────────┘",
]
`;

exports[`Array drawBoxInnerThickOuterThin [Variable Row Lengths 1+1+3] 1`] = `
[
  "┌─────────────────┰────────────┰─────────────┐",
  "│                 ┃Cell 2-1 ABC┃             │",
  "┝━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┥",
  "│Cell 1-2 Avatar  ┃            ┃             │",
  "┝━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┥",
  "│Cell 1-3 Lemmings┃Cell 2-3 A  ┃Cell 3-3 Bata│",
  "└─────────────────┸────────────┸─────────────┘",
]
`;

exports[`Array drawBoxInnerThickOuterThin [Variable Row Lengths 1+2] 1`] = `
[
  "┌────────┰────────┐",
  "│Cell 1-1┃        │",
  "┝━━━━━━━━╋━━━━━━━━┥",
  "│Cell 1-2┃Cell 2-2│",
  "└────────┸────────┘",
]
`;

exports[`Array drawBoxInnerThickOuterThinCurve [Simple 1X1] 1`] = `
[
  "╭────────────╮",
  "│In 1x1 array│",
  "╰────────────╯",
]
`;

exports[`Array drawBoxInnerThickOuterThinCurve [Simple 2x2] 1`] = `
[
  "╭────────┰────────╮",
  "│Cell 1-1┃Cell 2-1│",
  "┝━━━━━━━━╋━━━━━━━━┥",
  "│Cell 1-2┃Cell 2-2│",
  "╰────────┸────────╯",
]
`;

exports[`Array drawBoxInnerThickOuterThinCurve [Variable Row Lengths 1+1+3] 1`] = `
[
  "╭─────────────────┰────────────┰─────────────╮",
  "│                 ┃Cell 2-1 ABC┃             │",
  "┝━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┥",
  "│Cell 1-2 Avatar  ┃            ┃             │",
  "┝━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┥",
  "│Cell 1-3 Lemmings┃Cell 2-3 A  ┃Cell 3-3 Bata│",
  "╰─────────────────┸────────────┸─────────────╯",
]
`;

exports[`Array drawBoxInnerThickOuterThinCurve [Variable Row Lengths 1+2] 1`] = `
[
  "╭────────┰────────╮",
  "│Cell 1-1┃        │",
  "┝━━━━━━━━╋━━━━━━━━┥",
  "│Cell 1-2┃Cell 2-2│",
  "╰────────┸────────╯",
]
`;

exports[`Array drawBoxInnerThinOuterDouble [Simple 1X1] 1`] = `
[
  "╔════════════╗",
  "║In 1x1 array║",
  "╚════════════╝",
]
`;

exports[`Array drawBoxInnerThinOuterDouble [Simple 2x2] 1`] = `
[
  "╔════════╤════════╗",
  "║Cell 1-1│Cell 2-1║",
  "╟────────┼────────╢",
  "║Cell 1-2│Cell 2-2║",
  "╚════════╧════════╝",
]
`;

exports[`Array drawBoxInnerThinOuterDouble [Variable Row Lengths 1+1+3] 1`] = `
[
  "╔═════════════════╤════════════╤═════════════╗",
  "║                 │Cell 2-1 ABC│             ║",
  "╟─────────────────┼────────────┼─────────────╢",
  "║Cell 1-2 Avatar  │            │             ║",
  "╟─────────────────┼────────────┼─────────────╢",
  "║Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata║",
  "╚═════════════════╧════════════╧═════════════╝",
]
`;

exports[`Array drawBoxInnerThinOuterDouble [Variable Row Lengths 1+2] 1`] = `
[
  "╔════════╤════════╗",
  "║Cell 1-1│        ║",
  "╟────────┼────────╢",
  "║Cell 1-2│Cell 2-2║",
  "╚════════╧════════╝",
]
`;

exports[`Array drawBoxInnerThinOuterThick [Simple 1X1] 1`] = `
[
  "┏━━━━━━━━━━━━┓",
  "┃In 1x1 array┃",
  "┗━━━━━━━━━━━━┛",
]
`;

exports[`Array drawBoxInnerThinOuterThick [Simple 2x2] 1`] = `
[
  "┏━━━━━━━━┯━━━━━━━━┓",
  "┃Cell 1-1│Cell 2-1┃",
  "┠────────┼────────┨",
  "┃Cell 1-2│Cell 2-2┃",
  "┗━━━━━━━━┷━━━━━━━━┛",
]
`;

exports[`Array drawBoxInnerThinOuterThick [Variable Row Lengths 1+1+3] 1`] = `
[
  "┏━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━┓",
  "┃                 │Cell 2-1 ABC│             ┃",
  "┠─────────────────┼────────────┼─────────────┨",
  "┃Cell 1-2 Avatar  │            │             ┃",
  "┠─────────────────┼────────────┼─────────────┨",
  "┃Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata┃",
  "┗━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━┛",
]
`;

exports[`Array drawBoxInnerThinOuterThick [Variable Row Lengths 1+2] 1`] = `
[
  "┏━━━━━━━━┯━━━━━━━━┓",
  "┃Cell 1-1│        ┃",
  "┠────────┼────────┨",
  "┃Cell 1-2│Cell 2-2┃",
  "┗━━━━━━━━┷━━━━━━━━┛",
]
`;

exports[`Array drawBoxThick [Simple 1X1] 1`] = `
[
  "┏━━━━━━━━━━━━┓",
  "┃In 1x1 array┃",
  "┗━━━━━━━━━━━━┛",
]
`;

exports[`Array drawBoxThick [Simple 2x2] 1`] = `
[
  "┏━━━━━━━━┳━━━━━━━━┓",
  "┃Cell 1-1┃Cell 2-1┃",
  "┣━━━━━━━━╋━━━━━━━━┫",
  "┃Cell 1-2┃Cell 2-2┃",
  "┗━━━━━━━━┻━━━━━━━━┛",
]
`;

exports[`Array drawBoxThick [Variable Row Lengths 1+1+3] 1`] = `
[
  "┏━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━┓",
  "┃                 ┃Cell 2-1 ABC┃             ┃",
  "┣━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┫",
  "┃Cell 1-2 Avatar  ┃            ┃             ┃",
  "┣━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━━┫",
  "┃Cell 1-3 Lemmings┃Cell 2-3 A  ┃Cell 3-3 Bata┃",
  "┗━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━━━┛",
]
`;

exports[`Array drawBoxThick [Variable Row Lengths 1+2] 1`] = `
[
  "┏━━━━━━━━┳━━━━━━━━┓",
  "┃Cell 1-1┃        ┃",
  "┣━━━━━━━━╋━━━━━━━━┫",
  "┃Cell 1-2┃Cell 2-2┃",
  "┗━━━━━━━━┻━━━━━━━━┛",
]
`;

exports[`Array drawBoxThin [Simple 1X1] 1`] = `
[
  "┌────────────┐",
  "│In 1x1 array│",
  "└────────────┘",
]
`;

exports[`Array drawBoxThin [Simple 2x2] 1`] = `
[
  "┌────────┬────────┐",
  "│Cell 1-1│Cell 2-1│",
  "├────────┼────────┤",
  "│Cell 1-2│Cell 2-2│",
  "└────────┴────────┘",
]
`;

exports[`Array drawBoxThin [Variable Row Lengths 1+1+3] 1`] = `
[
  "┌─────────────────┬────────────┬─────────────┐",
  "│                 │Cell 2-1 ABC│             │",
  "├─────────────────┼────────────┼─────────────┤",
  "│Cell 1-2 Avatar  │            │             │",
  "├─────────────────┼────────────┼─────────────┤",
  "│Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata│",
  "└─────────────────┴────────────┴─────────────┘",
]
`;

exports[`Array drawBoxThin [Variable Row Lengths 1+2] 1`] = `
[
  "┌────────┬────────┐",
  "│Cell 1-1│        │",
  "├────────┼────────┤",
  "│Cell 1-2│Cell 2-2│",
  "└────────┴────────┘",
]
`;

exports[`Array drawBoxThinCurve [Simple 1X1] 1`] = `
[
  "╭────────────╮",
  "│In 1x1 array│",
  "╰────────────╯",
]
`;

exports[`Array drawBoxThinCurve [Simple 2x2] 1`] = `
[
  "╭────────┬────────╮",
  "│Cell 1-1│Cell 2-1│",
  "├────────┼────────┤",
  "│Cell 1-2│Cell 2-2│",
  "╰────────┴────────╯",
]
`;

exports[`Array drawBoxThinCurve [Variable Row Lengths 1+1+3] 1`] = `
[
  "╭─────────────────┬────────────┬─────────────╮",
  "│                 │Cell 2-1 ABC│             │",
  "├─────────────────┼────────────┼─────────────┤",
  "│Cell 1-2 Avatar  │            │             │",
  "├─────────────────┼────────────┼─────────────┤",
  "│Cell 1-3 Lemmings│Cell 2-3 A  │Cell 3-3 Bata│",
  "╰─────────────────┴────────────┴─────────────╯",
]
`;

exports[`Array drawBoxThinCurve [Variable Row Lengths 1+2] 1`] = `
[
  "╭────────┬────────╮",
  "│Cell 1-1│        │",
  "├────────┼────────┤",
  "│Cell 1-2│Cell 2-2│",
  "╰────────┴────────╯",
]
`;

exports[`Array drawBoxVerticalBarPlusEquals [Simple 1X1] 1`] = `
[
  "+============+",
  "|In 1x1 array|",
  "+============+",
]
`;

exports[`Array drawBoxVerticalBarPlusEquals [Simple 2x2] 1`] = `
[
  "+========+========+",
  "|Cell 1-1|Cell 2-1|",
  "+========+========+",
  "|Cell 1-2|Cell 2-2|",
  "+========+========+",
]
`;

exports[`Array drawBoxVerticalBarPlusEquals [Variable Row Lengths 1+1+3] 1`] = `
[
  "+=================+============+=============+",
  "|                 |Cell 2-1 ABC|             |",
  "+=================+============+=============+",
  "|Cell 1-2 Avatar  |            |             |",
  "+=================+============+=============+",
  "|Cell 1-3 Lemmings|Cell 2-3 A  |Cell 3-3 Bata|",
  "+=================+============+=============+",
]
`;

exports[`Array drawBoxVerticalBarPlusEquals [Variable Row Lengths 1+2] 1`] = `
[
  "+========+========+",
  "|Cell 1-1|        |",
  "+========+========+",
  "|Cell 1-2|Cell 2-2|",
  "+========+========+",
]
`;

exports[`Array drawBoxVerticalBarPlusMinus [Simple 1X1] 1`] = `
[
  "+------------+",
  "|In 1x1 array|",
  "+------------+",
]
`;

exports[`Array drawBoxVerticalBarPlusMinus [Simple 2x2] 1`] = `
[
  "+--------+--------+",
  "|Cell 1-1|Cell 2-1|",
  "+--------+--------+",
  "|Cell 1-2|Cell 2-2|",
  "+--------+--------+",
]
`;

exports[`Array drawBoxVerticalBarPlusMinus [Variable Row Lengths 1+1+3] 1`] = `
[
  "+-----------------+------------+-------------+",
  "|                 |Cell 2-1 ABC|             |",
  "+-----------------+------------+-------------+",
  "|Cell 1-2 Avatar  |            |             |",
  "+-----------------+------------+-------------+",
  "|Cell 1-3 Lemmings|Cell 2-3 A  |Cell 3-3 Bata|",
  "+-----------------+------------+-------------+",
]
`;

exports[`Array drawBoxVerticalBarPlusMinus [Variable Row Lengths 1+2] 1`] = `
[
  "+--------+--------+",
  "|Cell 1-1|        |",
  "+--------+--------+",
  "|Cell 1-2|Cell 2-2|",
  "+--------+--------+",
]
`;

exports[`Array drawBoxVerticalBarPlusMinusSlashes [Simple 1X1] 1`] = `
[
  "/------------\\",
  "|In 1x1 array|",
  "\\------------/",
]
`;

exports[`Array drawBoxVerticalBarPlusMinusSlashes [Simple 2x2] 1`] = `
[
  "/--------+--------\\",
  "|Cell 1-1|Cell 2-1|",
  "+--------+--------+",
  "|Cell 1-2|Cell 2-2|",
  "\\--------+--------/",
]
`;

exports[`Array drawBoxVerticalBarPlusMinusSlashes [Variable Row Lengths 1+1+3] 1`] = `
[
  "/-----------------+------------+-------------\\",
  "|                 |Cell 2-1 ABC|             |",
  "+-----------------+------------+-------------+",
  "|Cell 1-2 Avatar  |            |             |",
  "+-----------------+------------+-------------+",
  "|Cell 1-3 Lemmings|Cell 2-3 A  |Cell 3-3 Bata|",
  "\\-----------------+------------+-------------/",
]
`;

exports[`Array drawBoxVerticalBarPlusMinusSlashes [Variable Row Lengths 1+2] 1`] = `
[
  "/--------+--------\\",
  "|Cell 1-1|        |",
  "+--------+--------+",
  "|Cell 1-2|Cell 2-2|",
  "\\--------+--------/",
]
`;
